### PR TITLE
resource/ebs_snapshot: fix crash in case of manual deletion

### LINKED ISSUE #3460: Crash on aws_ebs_snapshot
_This issue was originally opened by @gdoumenc as hashicorp/terraform#17394. It was migrated here as a result of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original body of the issue is below._

<hr>

The reason of the crash was the fact that the snapshot was deleted from the aws console... I removed this resource in the state file (on S3) and everything was ok.

### Terraform Version
```
Terraform v0.11.3
+ provider.aws v1.7.1
```

### Terraform Configuration Files
```hcl
...
resource "aws_ebs_snapshot" "briteheart-odoo-for-canary" {
  volume_id = "${aws_ebs_volume.briteheart-odoo.id}"
  tags {
    Name = "briteheart-odoo-for-canary"
  }
}

```

### Crash Output
Error: Error refreshing state: 1 error(s) occurred:

* aws_ebs_snapshot.briteheart-odoo-for-canary: 1 error(s) occurred:

* aws_ebs_snapshot.briteheart-odoo-for-canary: aws_ebs_snapshot.briteheart-odoo-for-canary: unexpected EOF

panic: runtime error: index out of range
2018-02-20T20:17:58.512+0100 [DEBUG] plugin.terraform-provider-aws_v1.7.1_x4: 
2018-02-20T20:17:58.512+0100 [DEBUG] plugin.terraform-provider-aws_v1.7.1_x4: goroutine 16 [running]:
2018-02-20T20:17:58.512+0100 [DEBUG] plugin.terraform-provider-aws_v1.7.1_x4: github.com/terraform-providers/terraform-provider-aws/aws.resourceAwsEbsSnapshotRead(0xc42029c000, 0x2480f60, 0xc4201b6500, 0xc42029c000, 0x0)
2018-02-20T20:17:58.512+0100 [DEBUG] plugin.terraform-provider-aws_v1.7.1_x4: 	/opt/teamcity-agent/work/222ea50a1b4f75f4/src/github.com/terraform-providers/terraform-provider-aws/aws/resource_aws_ebs_snapshot.go:108 +0x64f
2018-02-20T20:17:58.512+0100 [DEBUG] plugin.terraform-provider-aws_v1.7.1_x4: github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/helper/schema.(*Resource).Refresh(0xc4203d3880, 0xc4205d20a0, 0x2480f60, 0xc4201b6500, 0xc42066d290, 0xc42015e901, 0x80000000018)
2018-02-20T20:17:58.512+0100 [DEBUG] plugin.terraform-provider-aws_v1.7.1_x4: 	/opt/teamcity-agent/work/222ea50a1b4f75f4/src/github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/helper/schema/resource.go:354 +0x17e
2018-02-20T20:17:58.513+0100 [DEBUG] plugin.terraform-provider-aws_v1.7.1_x4: github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/helper/schema.(*Provider).Refresh(0xc420866a80, 0xc4205d2050, 0xc4205d20a0, 0x7efd5cb47000, 0x0, 0x275)
2018-02-20T20:17:58.513+0100 [DEBUG] plugin.terraform-provider-aws_v1.7.1_x4: 	/opt/teamcity-agent/work/222ea50a1b4f75f4/src/github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/helper/schema/provider.go:308 +0x9a
2018-02-20T20:17:58.513+0100 [DEBUG] plugin.terraform-provider-aws_v1.7.1_x4: github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/plugin.(*ResourceProviderServer).Refresh(0xc420876040, 0xc4205ca030, 0xc4205ca190, 0x0, 0x0)
2018-02-20T20:17:58.513+0100 [DEBUG] plugin.terraform-provider-aws_v1.7.1_x4: 	/opt/teamcity-agent/work/222ea50a1b4f75f4/src/github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/plugin/resource_provider.go:549 +0x4e
2018-02-20T20:17:58.513+0100 [DEBUG] plugin.terraform-provider-aws_v1.7.1_x4: reflect.Value.call(0xc420054840, 0xc420684048, 0x13, 0x286746b, 0x4, 0xc420361f20, 0x3, 0x3, 0x0, 0x0, ...)
2018-02-20T20:17:58.513+0100 [DEBUG] plugin.terraform-provider-aws_v1.7.1_x4: 	/usr/local/go/src/reflect/value.go:434 +0x905
2018-02-20T20:17:58.513+0100 [DEBUG] plugin.terraform-provider-aws_v1.7.1_x4: reflect.Value.Call(0xc420054840, 0xc420684048, 0x13, 0xc420386f20, 0x3, 0x3, 0x0, 0x0, 0x0)
2018-02-20T20:17:58.513+0100 [DEBUG] plugin.terraform-provider-aws_v1.7.1_x4: 	/usr/local/go/src/reflect/value.go:302 +0xa4
2018-02-20T20:17:58.513+0100 [DEBUG] plugin.terraform-provider-aws_v1.7.1_x4: net/rpc.(*service).call(0xc420170500, 0xc420682000, 0xc4204be060, 0xc420686480, 0xc42016c280, 0x200c420, 0xc4205ca030, 0x16, 0x200c460, 0xc4205ca190, ...)
2018-02-20T20:17:58.513+0100 [DEBUG] plugin.terraform-provider-aws_v1.7.1_x4: 	/usr/local/go/src/net/rpc/server.go:381 +0x142
2018-02-20T20:17:58.513+0100 [DEBUG] plugin.terraform-provider-aws_v1.7.1_x4: created by net/rpc.(*Server).ServeCodec
2018-02-20T20:17:58.513+0100 [DEBUG] plugin.terraform-provider-aws_v1.7.1_x4: 	/usr/local/go/src/net/rpc/server.go:475 +0x36b
2018-02-20T20:17:58.517+0100 [DEBUG] plugin: plugin process exited: path=/home/studiogdo/workspace/briteheart/infra/terraform/.terraform/plugins/linux_amd64/terraform-provider-aws_v1.7.1_x4
2018/02/20 20:17:58 [ERROR] root: eval: *terraform.EvalRefresh, err: aws_ebs_snapshot.briteheart-odoo-for-canary: unexpected EOF
2018/02/20 20:17:58 [ERROR] root: eval: *terraform.EvalSequence, err: aws_ebs_snapshot.briteheart-odoo-for-canary: unexpected EOF
2018/02/20 20:17:58 [TRACE] [walkRefresh] Exiting eval tree: aws_ebs_snapshot.briteheart-odoo-for-canary
2018/02/20 20:17:58 [TRACE] dag/walk: upstream errored, not walking "provider.aws (close)"
2018/02/20 20:17:58 [TRACE] dag/walk: upstream errored, not walking "root"
2018/02/20 20:17:58 [DEBUG] plugin: waiting for all plugin processes to complete...
2018-02-20T20:17:58.518+0100 [WARN ] plugin: error closing client during Kill: err="connection is shut down"

!!!!!!!!!!!!!!!!!!!!!!!!!!! TERRAFORM CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!

Terraform crashed! This is always indicative of a bug within Terraform.
A crash log has been placed at "crash.log" relative to your current
working directory. It would be immensely helpful if you could please
report the crash with Terraform[1] so that we can fix this.

When reporting bugs, please include your terraform version. That
information is available on the first line of crash.log. You can also
get it by running 'terraform --version' on the command line.

[1]: https://github.com/hashicorp/terraform/issues

### PR DESCRIPTION
**Pull request by loivis**
Fix #3460 

Also enhance error checking in case of any other error.

### CODE REVIEW

**Comment by bflad** [comment]
The error code change LGTM according to the [EC2 API documentation](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/errors-overview.html#api-error-codes-table-client).

Can you add an acceptance test that fails then passes to prevent this in the future and verify your fix? e.g.
```go
{
  Config: //basic
}
{
  PreConfig: func() {
    // SDK call to DeleteSnapshot
  },
  Config: // basic
}
```

We may also want to have an additional check to remove the ID from the state and return nil if `len(res.Snapshots) == 0`

File: aws/resource_aws_ebs_snapshot.go [hunk 1]
```diff
@@ -99,10 +99,13 @@ func resourceAwsEbsSnapshotRead(d *schema.ResourceData, meta interface{}) error
 		SnapshotIds: []*string{aws.String(d.Id())},
 	}
 	res, err := conn.DescribeSnapshots(req)
-	if ec2err, ok := err.(awserr.Error); ok && ec2err.Code() == "InvalidSnapshotID.NotFound" {
-		log.Printf("Snapshot %q Not found - removing from state", d.Id())
-		d.SetId("")
-		return nil
+	if err != nil {
+		if isAWSErr(err, "InvalidSnapshot.NotFound", "") {
+			log.Printf("Snapshot %q Not found - removing from state", d.Id())
```
Comment (radeksimko) [inline aws/resource_aws_ebs_snapshot.go]: Nitpick: We tend to prefix such messages with `[WARN]` so it's easy to find them in the log.

Comment (loivis) [reply aws/resource_aws_ebs_snapshot.go]: of coz

**Comment by loivis** [comment]
anything else to change before merge?

File: aws/resource_aws_ebs_snapshot_test.go [hunk 2]
```diff
@@ -91,58 +129,71 @@ func testAccCheckSnapshotExists(n string, v *ec2.Snapshot) resource.TestCheckFun
...
 
   tags {
-    Name = "testAccAwsEbsSnapshotConfigWithKms"
+    Name = "${var.name}"
   }
 }
 
 resource "aws_ebs_volume" "test" {
-  availability_zone = "us-west-2a"
+  availability_zone = "${data.aws_region.current.name}a"
```
Comment (bflad) [inline aws/resource_aws_ebs_snapshot_test.go]: Minor nitpick: Not all regions have/will have AZ `a` available but this is certainly an improvement (`data.aws_availability_zones.current[0].id` is safer)

**Comment by bflad** [comment]
This has been released in [version 1.11.0 of the AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#1110-march-09-2018). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.

**Pull request merged**